### PR TITLE
Use vim.empty_dict() instead of {}

### PR DESCRIPTION
**Pull request by h-michael**
fix: #94
rust-analyzer parses initializeOptions according to the type definition.
Therefore, an empty dictionary fails because it does not satisfy the type definition.
So, set the default value of ServerConfig type to initializeOptions.

https://github.com/rust-analyzer/rust-analyzer/blob/d8d8c20077702b8537086d49914d02654a46ebc5/crates/ra_lsp_server/src/main.rs#L65-L79
https://github.com/rust-analyzer/rust-analyzer/blob/dc48f89581843248660ceb755bb20469ab6ac0c9/crates/ra_lsp_server/src/config.rs#L18-L49

### CODE REVIEW

**Comment by justinmk** [comment]
That seems brittle, and defeats the purpose of loose formats like JSON. @matklad is this expected behavior of rust-analyzer: that clients have to send all fields in init_options?

**Comment by h-michael** [comment]
@justinmk initializeOptions keis optional and value is any type.
So how to handle this option is different from each servers.
So we might remove initializeOptions key if value is empty dict or nil.

**Comment by matklad** [comment]
Nope, that would be a bug. We intend to handle empty object correctly, we use serde default:

```
#[serde(rename_all = "camelCase", default)]
```

But #94 says that we see an empty **array** here, and not an empty object, and that shape mismatch is I think worth complaining about?

**Comment by matklad** [comment]
Anyway, I think it makes sense to double check on our side how do we handle empty objects, and to improve the error message to include the original json, thanks for the ping!

**Comment by justinmk** [comment]
> But #94 says that we see an empty **array** here, and not an empty object, and that shape mismatch is I think worth complaining about?

Yes, we are sending an object instead of array since https://github.com/neovim/nvim-lsp/commit/b4f4b72a50983464864810e642e0b70e0b987e6f . 

> Anyway, I think it makes sense to double check on our side how do we handle empty objects, and to improve the error message to include the original json, thanks for the ping!

Thanks!

**Comment by matklad** [comment]
I've just checked, we do handle `{}`. Here's log from the corresponding VS Code invocation: https://gist.github.com/matklad/86adaeb6a52feb87cb5ceb850ed14810.

I've also pushed https://github.com/rust-analyzer/rust-analyzer/pull/3102 which should include the original JSON in the error message, which should help with debugging the issue.

**Comment by h-michael** [comment]
@matklad Thank you for your quickly fixing. :)

**Comment by matklad** [comment]
@h-michael could you try wit this new version of rust-analyzer and paste the error message you get? I hope it sheds the light on the issue.

**Comment by h-michael** [comment]
@matklad 
Sorry, I misunderstood.
This error didn't seem to be the response of initialize.
https://gist.github.com/h-michael/deb0052a00456709bcb2f557c94bf083

**Comment by justinmk** [comment]
> "[ERROR ra_lsp_server] Failed to deserialize config: invalid length 0, expected struct ServerConfig with 12 elements; []\n"

That [] suggests that somewhere we're still sending an array ?

File: lua/nvim_lsp/rust_analyzer.lua [hunk 1]
```diff
@@ -8,7 +8,23 @@ configs.rust_analyzer = {
     filetypes = {"rust"};
     root_dir = util.root_pattern("Cargo.toml");
     log_level = lsp.protocol.MessageType.Warning;
-    settings = {};
```
Comment (justinmk) [inline lua/nvim_lsp/rust_analyzer.lua]: does this help?

    settings = vim.empty_dict();

**Comment by h-michael** [comment]
I think https://github.com/neovim/neovim/pull/11864 will solve this.

File: lua/nvim_lsp/configs.lua [hunk 2]
```diff
@@ -94,7 +94,7 @@ function configs.__newindex(t, config_name, config_def)
       M.manager = nil
     end
     local manager = util.server_per_root_dir_manager(function(_root_dir)
-      local new_config = vim.tbl_extend("keep", {}, config)
+      local new_config = vim.tbl_extend("keep", vim.empty_dict(), config)
```
Comment (justinmk) [inline lua/nvim_lsp/configs.lua]: Nice catch!

**Comment by h-michael** [comment]
@matklad That was our problem. Thank you for helping us debug.

**Pull request merged**